core/src/test/java/org/springframework/security/provisioning/InMemoryUserDetailsManagerTests.java

```diff
@@ -97,4 +102,88 @@ public void changePasswordWhenCustomSecurityContextHolderStrategyThenUses() {
 		verify(strategy).getContext();
 	}
 
+	@Test
+	public void createUserWhenUserAlreadyExistsThenException() {
+		assertThatIllegalArgumentException().isThrownBy(() -> this.manager.createUser(this.user))
+			.withMessage("user should not exist");
+	}
+
+	@Test
+	public void createUserWhenNotInstanceOfMutableUserDetailsThenShouldWrapIntoMutableUser() {
+		InMemoryUserDetailsManager manager = new InMemoryUserDetailsManager();
+		final UserDetails user = User.withUserDetails(PasswordEncodedUser.user()).build();
+		manager.createUser(user);
+		UserDetails userDetails = manager.loadUserByUsername(user.getUsername());
+		assertThat(userDetails).isInstanceOf(User.class);
+	}
+
+	@Test
+	public void createUserWhenInstanceOfMutableUserDetailsThenShouldNotWrapIntoMutableUser() {
+		InMemoryUserDetailsManager manager = new InMemoryUserDetailsManager();
+		CustomUser user = new CustomUser(User.withUserDetails(PasswordEncodedUser.user()).build());
+		manager.createUser(user);
+		UserDetails userDetails = manager.loadUserByUsername(user.getUsername());
+		assertThat(userDetails).isInstanceOf(CustomUser.class);
+	}
+
+	@Test
+	public void loadUserByUsernameWhenUserNullThenException() {
```

Comment:
Thank you for the extra tests. Will you please place them in a separate commit so that I can backport those to earlier versions?

Reply:
I have placed them in a separate commit.